File: 6. Zigzag Conversion/zigzag-conversion.py
def convert(s, numRows):
    if numRows == 1 or numRows >= len(s):
        return s

    rows = ["" for _ in range(numRows)]
    curr_row = 0
    direction = -1

    for char in s:
        rows[curr_row] += char

        if curr_row == 0 or curr_row == numRows - 1:
            direction *= -1
        curr_row += direction

    return "".join(rows)

if __name__ == "__main__":
    print(convert("PAYPALISHIRING", 3))  # "PAHNAPLSIIGYIR"
    print(convert("PAYPALISHIRING", 4))  # "PINALSIGYAHRPI"
    print(convert("A", 1))               # "A"

File: 6. Zigzag Conversion/zigzag-conversion.js
function convert(s, numRows) {
    if (numRows === 1 || numRows >= s.length) {
        return s;
    }

    const rows = new Array(numRows).fill("");
    let currRow = 0;
    let direction = -1;

    for (const char of s) {
        rows[currRow] += char;
        if (currRow === 0 || currRow === numRows - 1) {
            direction *= -1;
        }
        currRow += direction;
    }

    return rows.join("");
}

function main() {
    console.log(convert("PAYPALISHIRING", 3)); // "PAHNAPLSIIGYIR"
    console.log(convert("PAYPALISHIRING", 4)); // "PINALSIGYAHRPI"
    console.log(convert("A", 1));              // "A"
}

main();
